core/services/head_tracker.go

```diff
@@ -234,27 +225,33 @@ func (ht *HeadTracker) receiveHeaders() error {
 }
 
 func (ht *HeadTracker) subscribeToHead() error {
+	ht.headMutex.Lock()
+	defer ht.headMutex.Unlock()
+
 	ht.headers = make(chan models.BlockHeader)
 	sub, err := ht.store.TxManager.SubscribeToNewHeads(ht.headers)
 	if err != nil {
 		return err
 	}
 	ht.headSubscription = sub
-	ht.connected.Set()
-	ht.connect(ht.Head())
+	ht.connect(ht.head)
+	ht.connected = true
```

Comment:
I originally set the connected bool to true before invoking callbacks because I felt it was more correct. If a callback asked `ht.Connected()` inside the callback, it should return true. Seems like this is now a non-issue with the change of passing callbacks into constructors, can we switch it back? It's still behind the mutex.

Reply:
Yep, I'll change this.